File: app/src/main/java/com/genius/wasylews/readme/domain/reader/BookReader.java
package com.genius.wasylews.readme.domain.reader;

import com.genius.wasylews.readme.domain.model.Book;
import com.genius.wasylews.readme.domain.model.BookPage;

import java.util.Iterator;

public class BookReader implements Iterator<BookPage> {

    private Book book;
    private int page = 0;

    public BookReader(Book book) {
        this.book = book;
    }

    @Override
    public boolean hasNext() {
        return page < book.getPageCount() - 1;
    }

    @Override
    public BookPage next() {
        return new BookPage(page);
    }
}

File: app/src/main/java/com/genius/wasylews/readme/domain/reader/BookReader.kt
package com.genius.wasylews.readme.domain.reader

import com.genius.wasylews.readme.domain.model.Book
import com.genius.wasylews.readme.domain.model.BookPage

class BookReader(private val book: Book) : ListIterator<BookPage> {
    private val page = 0

    override fun hasNext(): Boolean {
        return page < book.pageCount - 1
    }

    override fun next(): BookPage {
        return BookPage(page)
    }

    override fun nextIndex(): Int {
        return page + 1
    }

    override fun hasPrevious(): Boolean {
        return page > 0
    }

    override fun previous(): BookPage {
        return BookPage(page)
    }

    override fun previousIndex(): Int {
        return page - 1
    }
}